products/epinio/edit/services.vue

```diff
@@ -77,21 +80,67 @@ export default Vue.extend<Data, any, any, any>({
         saveCb(false);
       }
     },
-    setValid(field: string, valid: boolean) {
-      this.validFields[field] = valid;
+    updateNameErrors(): void {
+      this.touchedName = true;
+      const nameErrors = this.getNameErrors(this.value.metadata.name || '');
+      const nameErrorMessage = nameErrors.join(', ');
+
+      this.nameErrorMessage = nameErrorMessage;
     },
-    meetsNameRequirements( name = '') {
-      const nameErrors = validateKubernetesName(name, this.t('epinio.namespace.name'), this.$store.getters, undefined, []);
-
-      if (nameErrors.length > 0) {
-        return {
-          isValid:      false,
-          errorMessage: nameErrors.join(', ')
-        };
+    updateDataErrors(): void {
+      this.touchedData = true;
+
+      const keys = Object.keys(this.value.data || {});
+      const values = Object.values(this.value.data || {});
+
+      if (keys.length === 0) {
+        const atLeastOnePairRequired = this.t(
+          'epinio.services.pairs.requirement'
+        );
+
+        this.dataErrorMessage = atLeastOnePairRequired;
+
+        return;
+      }
+      if (keys.includes('') || values.includes('')) {
+        const cannotBeEmpty = this.t('epinio.services.pairs.empty');
+
+        this.dataErrorMessage = cannotBeEmpty;
       }
+      const regex = /^[a-zA-Z0-9_\-\.]*$/gm;
+      const keyHasInvalidChars = this.t('epinio.services.pairs.characters');
 
-      return { isValid: true };
+      for (let i = 0; i < keys.length; i++) {
+        const key = keys[i];
+
+        if (!key.match(regex)) {
+          this.dataErrorMessage = keyHasInvalidChars;
+
+          return;
+        }
+      }
+      this.dataErrorMessage = '';
     },
+    getNameErrors(name: string): string[] {
+      return validateKubernetesName(
+        this.value.metadata.name,
+        this.t('epinio.namespace.name'),
+        this.$store.getters,
+        undefined,
+        []
+      );
+    },
+  },
+  watch: {
+    'value.data': {
+      // We need this deep watcher so that the
+      // error messages for the key-value pairs in data
+      // are updated on every keystroke.
+      handler() {
+        this.updateDataErrors();
```

Comment:
Isn't this handled by the updateErrors process, could that be used to handle the `remove/add/read from file` process as well?

Reply:
updateErrors is just the name of the event. The reason I called the event `updateErrors` instead of `updateDataErrors` is because I was just avoiding a reference to epinio data in the more generic components.